bootstrap/kubeadm/internal/cloudinit/controlplane_init.go

```diff
@@ -31,9 +31,14 @@ const (
 {{.ClusterConfiguration | Indent 6}}
       ---
 {{.InitConfiguration | Indent 6}}
+-   path: /run/cluster-api/bootstrap
+    owner: root:root
+    permissions: '0640'
+    content: ""
 runcmd:
 {{- template "commands" .PreKubeadmCommands }}
   - 'kubeadm init --config /run/kubeadm/kubeadm.yaml {{.KubeadmVerbosity}}'
+  - {{ .SentinelFileCommand }}
```

Comment:
Wondering if this should go here or after the PostKubeadmCommands.

Here we are documenting kubeadm init has completed, which is usually the most critical part of a bootstrap process, but I assume that from a user PoV also the PostKubeadmCommands are an integral part of the same sequence as well.

Another option could be to have more than one sentinel files, but I would avoid such complications if not strictly necessary.

Reply:
> Wondering if this should go here or after the PostKubeadmCommands.

Disagree. The user is in control of the "post" commands and is able to properly detect/handle errors on their own. The primary (and only?) goal here is to detect failure modes of "kubeadm init".

That said, I'm not convinced that this PR achieves this. In local testing, I've observed that cloudinit doesn't necessarily abort if one of the "pre" or "post" commands fail. AFAICT it has to do with how "string" runcmd values are wrapped (internally, by cloudinit) by an "sh" command:

https://cloudinit.readthedocs.io/en/latest/topics/modules.html?highlight=runcmd#runcmd

^ this doc does not actually describe the failure mode when "runcmd" value is a list of strings, i.e. whether or not all commands are always executed, or else an error in the flow aborts the remaining commands. so the wrapping "sh" may or may not be executed as if invoked with a `-e` shell option. for consistent behavior, i've written my own generator code to consolidate multiple "pre" commands into a bash script whose header includes `set -eu; set -o pipefail;` in order to get abortive behavior.

why is this important? well, if `kubeadm init ...` actually fails, but cloudinit doesn't abort and simply executes the "sentinel-file-command" anyway, then this part of the PR doesn't add much value. what's actually need is the equivalent of:
```yaml
runcmd:
- ... pre-commands ...
- [ bash -e -c "kubeadm init ... ; {{ .SentinelFileCommand }}" ] # cloudinit doesn't wrap this with `sh` because it's a list, it's executed via `execve` instead (note the "-e" flag passed to the shell)
- ... post-commands ...
```

Reply:
Also, because this is subtle, I wonder if it's a good idea to have some kind of integration test in this PR to guard against regressions.

Reply:
> Disagree. The user is in control of the "post" commands and is able to properly detect/handle errors on their own. The primary (and only?) goal here is to detect failure modes of "kubeadm init".

+1, if the user wants to detect errors in postKubeadmCommands, they can implement that themselves. The contract for this file is to signal success of Kubernetes node bootstrapping. 

> why is this important? well, if kubeadm init ... actually fails, but cloudinit doesn't abort and simply executes the "sentinel-file-command" anyway, then this part of the PR doesn't add much value. what's actually need is the equivalent of:

At first I had `kubeadm init && {{ .SentinelFileCommand }}`. The issue with that (and same thing with `kubeadm init;  {{ .SentinelFileCommand }}"` is that we do some string manipulation in the Docker provider to append `--ignore-preflight-errors=all` to the end of the string that contains "kubeadm init" or "kubeadm join" so we end up with `kubeadm init && {{ .SentinelFileCommand }} --ignore-preflight-errors=all` which fails... That said, it is essential that the sentinel file is only writen if and only if kubeadm init/join succeeds, since the point is to report successful completion. We might have to modify the Docker provider to not make assumptions on the format of the runcmd strings. @fabriziopandini thoughts?

Reply:
@CecileRobertMichon @jdef thanks for the context for postKubeadmCommands, might be let's add a comment in the API to make this more evident and documented.

WRT to changing the fake cloud-init implemented in the docker provider, I don't see problems in making changes